### DISCUSSION TITLE
Enhance debugging experience with new VSCode Debug Visualizer API

### DISCUSSION THREAD

**Discussion by NickAcPT** [rust-lang/rust-analyzer#16717] [Ideas]
Hey there,

With the release of VSCode 1.87, a [Debug Visualizer API](https://github.com/microsoft/vscode/blob/15a0f3c48f9315669c97f101a5d954464285e6ac/src/vscode-dts/vscode.proposed.debugVisualization.d.ts) has been proposed.

Per the [update log](https://code.visualstudio.com/updates/v1_87#_debug-visualizer-api):
> We have a new API that enables extensions to contribute _visualizers_ for a variable. These visualizers might be either actions that run a command, such as opening a new editor, or trees embedded inline into the Debug views and that take the place of data's default representation.

Although I'm painfully aware that rust-analyzer "has nothing to do with debugging" (https://github.com/rust-lang/rust-analyzer/issues/15407), I firmly believe that we could harness rust-analyzer's knowledge of the codebase (memory layouts; etc) to provide a slightly better debugging experience.

Extra links:
- VSCode's own issue for debug visualization: https://github.com/microsoft/vscode/issues/197287

**Comment by lnicola** [comment]
> I firmly believe that we could harness rust-analyzer's knowledge of the codebase (memory layouts; etc) to provide a slightly better debugging experience.

Our knowledge of the memory layout is somewhat unreliable and pretty much stops there. Even if it was accurate, it wouldn't apply to optimized binaries. Only the compiler and a debugger have a chance to understand it, and you can count the production-ready debuggers on the fingers on one hand.

Sure, it would be nice to view an `image::RgbImage` in VS Code, but I don't see that happening.

**Reply by NickAcPT** [reply]
> Sure, it would be nice to view an `image::RgbImage` in VS Code, but I don't see that happening.

When I was writing this suggestion, I wasn't even thinking about _that_, although that's an interesting idea.

A common frustration of mine has been with Enums when debugging where it's just hard to see what things are, but I guess that's still way out of r-a's scope even with this API being introduced 😅

**Reply by lnicola** [reply]
![image](https://github.com/rust-lang/rust-analyzer/assets/308347/0d20bef0-9d3c-428c-95e4-1983316e8826)

![image](https://github.com/rust-lang/rust-analyzer/assets/308347/4ebcc959-40e8-4874-86a4-709771c297e1)

I have no idea :shrug:.

**Reply by lnicola** [reply]
At least `rust-gdb` works:

![image](https://github.com/rust-lang/rust-analyzer/assets/308347/8ef52f83-e466-42f8-8209-3dd35ac77c03)

**Reply by lnicola** [reply]
And, despite some prodding, Native Debug doesn't work too well:

![image](https://github.com/rust-lang/rust-analyzer/assets/308347/4f299bc7-a784-4edf-91b5-7a56f8046f5f)

**Reply by lnicola** [reply]
And this is C/C++, again with some prodding:

![image](https://github.com/rust-lang/rust-analyzer/assets/308347/6a00e93e-b6f4-40c7-a6d2-8937f443ae30)

Pretty sad TBH.
